SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialComponentTest/SpatialComponentTest.cpp

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialComponentTest.h"
+#include "Engine/World.h"
+#include "Net/UnrealNetwork.h"
+#include "SpatialComponentTestActor.h"
+#include "SpatialComponentTestCallbackComponent.h"
+#include "SpatialComponentTestDummyComponent.h"
+#include "SpatialComponentTestReplicatedActor.h"
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialGDK/Public/EngineClasses/SpatialNetDriver.h"
+#include "SpatialGDK/Public/Utils/SpatialStatics.h"
+
+/** This Test is meant to check that you can add / remove components on spatial component callbacks for the following cases:
+ *		- replicated level actor
+ *		- dynamic replicated actor (spawned on the same server)
+ *		- dynamic replicated actor (spawned cross server)
+ * Keep in mind that we're assuming a 2x1 Grid Load-Balancing Strategy, otherwise the ownership of
+ * these actors may be something completely different (specially important for actors placed in the Level).
+ */
+ASpatialComponentTest::ASpatialComponentTest()
+{
+	Author = "Victoria Bloom";
+	Description = TEXT("Test GDK component callbacks");
+
+	Server1Position = FVector(-250.0f, -250.0f, 0.0f);
+	Server2Position = FVector(250.0f, -250.0f, 0.0f);
+}
+
+void ASpatialComponentTest::PrepareTest()
+{
+	Super::PrepareTest();
+
+	// Reusable functional test steps
+
+	FSpatialFunctionalTestStepDefinition ReplicatedDestroyStepDefinition(/*bIsNativeDefinition*/ true);
+	ReplicatedDestroyStepDefinition.StepName = TEXT("Replicated Dynamic Actor Spawned On Different Server - Destroy");
+	ReplicatedDestroyStepDefinition.TimeLimit = 5.0f;
+	ReplicatedDestroyStepDefinition.NativeStartEvent.BindLambda([this]() {
+		// Destroy to be able to re-run.
+		DynamicReplicatedActor->Destroy();
+		CrossServerSetDynamicReplicatedActor(nullptr);
+		FinishStep();
+	});
+
+	ResetTimer();
+
+	// Replicated Level Actor. Server 1 should have Authority, again assuming that the Level is setup accordingly.
+	{
+		AddStep(
+			TEXT("Replicated Level Actor - Verify Server Components"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+			[this](float DeltaTime) {
+				Timer -= DeltaTime;
+				if (Timer <= 0)
+				{
+					CheckComponents(LevelReplicatedActor, 1);
+				}
+			},
+			5.0f);
+	}
+
+	// Replicated Dynamic Actor Spawned On Same Server.Server 1 should have Authority.
+	{
+		AddStep(TEXT("Replicated Dynamic Actor Spawned On Same Server - Spawn"), FWorkerDefinition::Server(1), nullptr, [this]() {
+			ASpatialComponentTestReplicatedActor* Actor =
+				GetWorld()->SpawnActor<ASpatialComponentTestReplicatedActor>(Server1Position, FRotator::ZeroRotator);
+			CrossServerSetDynamicReplicatedActor(Actor);
+			FinishStep();
+		});
+
+		AddStep(
+			TEXT("Replicated Dynamic Actor Spawned On Same Server - Verify components"), FWorkerDefinition::AllWorkers, nullptr, nullptr,
+			[this](float DeltaTime) {
```

Comment:
could this be done with RequireTrue instead of timers?

Reply:
Will use RequireTrue instead of if checks in the CheckComponents. The timers were just to wait before checking but not necessary so I have removed them.